docs/designs/checkpoint_design.md

```diff
@@ -0,0 +1,59 @@
+
+## Design Components
+There are two design components: export and restore.
+
+### Export Model Parameters to a Checkpoint
+
+**When to save a checkpoint during training.** 
+
+In ElasticDL, we will save all model parameters to a checkpoint directory every `checkpoint_steps`. Though, some frameworks support to save checkpoints every N steps or every N seconds, e.g. `tf.estimator`. Using ParameterServerStrategy in ElasticDL, PS instances are asynchronous and we cannot guarantee that the iteration steps of all PS instances are consistent at the same time. So, we don't support to save checkpoints every N seconds.
+
+
+**Where to save a checkpoint.** 
+
+We need to set the `checkpoint_dir` in args if we want to save checkpoints. The saved checkpoint directory will be generated by `checkpoint_dir` and iteration steps which we call model version in ElasticDL, e.g. `/{checkpoint_dir}/model_v{version}/`.
+```python
+checkpoint_folder_name = "model_v%s" % version
+checkpoint_dir = os.path.join(
+    args.checkpoint_dir, checkpoint_folder_name
+)
+```
+ElasticDL is a Kubernetes-native framework and local directory in the pod will be lost if the pod exits in Kubernetes. So the `checkpoint_dir` should be a persistent directory which will not be lost after the pod exit. Now, we can utilize `hostPath` and `PersistentVolume` to get a persistent directory to save checkpoints in ElasticDL.
+
+**Save model parameters to a checkpoint directory.**
+
+Using AllReduceStrategy in ElasticDL, each worker has all parameters and we can assign any worker to save all parameters to a checkpoint directory. Using ParameterServerStrategy, all model parameters are partitioned on all PS instances. They need to save their owner parameters to the same checkpoint directory at the same step. Saving checkpoints for AllReduceStrategy can be seen as a special case that there is only a PS instance using ParameterServerStrategy. So, we can adopt the same way to save checkpoints for AllReduceStrategy and ParameterServerStrategy.
+
+The parameters on each PS instance may contain non-embedding variables and a part of embedding vectors of each embedding table. We create a `Tensor` protobuf with values for each variable or with values and indices for embedding vectors. To save an embedding table, we also need to create `EmbeddingTableInfo` to save the meta information like `initializer` and `dim`. Then, we create a `Model` protobuf including those `Tensor`s and `EmbeddingTableInfo`s. Finally, we save the `Model` to a  file named `variables-{ps_id}-of-{ps_num}.chkpt` where `ps_id` is the index of a PS instance and `ps_num` is the total number of PS instances. The checkpoint file will be located in a subdirectory `model_v{version}` for each checkpoint version. After the iteration versions of all PS instances exceed the checkpoint version in `model_v{version}`,  the `model_v{version}` will contain files of the entire model parameters.
+
+**The maximum number of recent checkpoints to keep**
```

Comment:
How many checkpoints to keep?

Reply:
Good suggestion. I will rename the 3rd title to "How to save model parameters to a checkpoint directory" in the section.